File: Source/Entities/TriggersAndRegions/PositionTrackerRegion.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Celeste.Mod.Entities;
using Celeste.Mod.Microlith57Misc.Components;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.Microlith57Misc.Entities;

[CustomEntity(
    "Microlith57Misc/PositionTrackerRegion=CreateFlag",
    "Microlith57Misc/PositionTrackerRegion_Expression=CreateExpr"
)]
public sealed class PositionTrackerRegion : Entity {

    #region --- State ---

    private enum TargetType {
        Player,
        Actor,
        NonPlayerActor,
        Solid,
    }

    private enum DetectionType {
        Within,
        Intersecting,
        Nearest,
    }

    private enum StickinessType {
        Free,
        Transient,
        UntilNewMatch,
        UntilDeath,
        Lifelink,
        Soulbond,
    }

    private enum TrackingType {
        Position,
        Center,
        TopCenter,
        BottomCenter,
        CenterLeft,
        CenterRight,
        Size,
    }

    private readonly ConditionSource Condition;
    public bool Enabled => Condition.Value;

    private TargetType WillTarget;
    private DetectionType Detection;
    private StickinessType Stickiness;
    private TrackingType Tracking;

    private Entity? Target;
    private bool EverTargetted = false;

    private Session.Slider SliderX, SliderY;
    private string TargettingFlag;

    #endregion State
    #region --- Init ---

    public PositionTrackerRegion(
        EntityData data, Vector2 offset,
        Session.Slider sliderX, Session.Slider sliderY,
        ConditionSource condition
    ) : base(data.Position + offset) {
        this.SetDepthAndTags(data);

        Collider = new Hitbox(data.Width, data.Height);
        Collidable = false;

        Add(Condition = condition);

        WillTarget = data.Enum("target", TargetType.Player);
        Detection = data.Enum("detection", DetectionType.Intersecting);
        Stickiness = data.Enum("stickiness", StickinessType.Lifelink);
        Tracking = data.Enum("tracking", TrackingType.Position);

        SliderX = sliderX;
        SliderY = sliderY;
        TargettingFlag = data.Attr("targettingFlag", "");
    }


    public static PositionTrackerRegion CreateFlag(Level level, LevelData _, Vector2 offset, EntityData data)
        => new(
            data, offset,
            level.Session.GetSliderObject(data.Attr("sliderPrefix", "trackedPosition") + 'X'),
            level.Session.GetSliderObject(data.Attr("sliderPrefix", "trackedPosition") + 'Y'),
            new ConditionSource.Flag(data, "retargetIfFlag", invertName: "invertRetargetIfFlag") { Default = true }
        );

    public static PositionTrackerRegion CreateExpr(Level level, LevelData _, Vector2 offset, EntityData data)
        => new(
            data, offset,
            level.Session.GetSliderObject(data.Attr("sliderPrefix", "trackedPosition") + 'X'),
            level.Session.GetSliderObject(data.Attr("sliderPrefix", "trackedPosition") + 'Y'),
            new ConditionSource.Expr(data, "retargetIfExpression") { Default = true }
        );

    #endregion Init
    #region --- Behaviour ---

    public override void Awake(Scene scene) {
        base.Awake(scene);
        UpdateTarget();
        UpdateOutputs();
    }

    public override void Update() {
        base.Update();
        UpdateTarget();
        UpdateOutputs();
    }

    private void UpdateOutputs() {
        if (Value is Vector2 vec) {
            SliderX.Value = vec.X;
            SliderY.Value = vec.Y;
        }

        if (Scene is Level level && TargettingFlag != "")
            level.Session.SetFlag(TargettingFlag, Target != null);
    }

    private Vector2? Value {
        get {
            if (Target == null) return null;
            switch (Tracking) {
                case TrackingType.Position: return Target.Position;
                case TrackingType.Center: return Target.Center;
                case TrackingType.TopCenter: return Target.TopCenter;
                case TrackingType.BottomCenter: return Target.BottomCenter;
                case TrackingType.CenterLeft: return Target.CenterLeft;
                case TrackingType.CenterRight: return Target.CenterRight;
                case TrackingType.Size:
                    return new(Target.Collider?.Width ?? 0f, Target.Collider?.Height ?? 0f);
                default: throw new Exception("unreachable");
            }
        }
    }

    #endregion Behaviour
    #region --- Targetting ---

    private void UpdateTarget() {
        bool canSwitch = Enabled;

        if (Target?.Scene == null)
            Target = null;

        Entity? target = null;

        Collidable = true;
        switch (Stickiness) {
            case StickinessType.Free:
                if (TryGetTarget(ref target) && canSwitch)
                    Target = target;
                goto case StickinessType.Transient;

            case StickinessType.Transient:
                if (Target != null && !Matches(Target) && canSwitch)
                    Target = null;
                goto case StickinessType.UntilDeath;

            case StickinessType.UntilNewMatch:
                if ((Target == null || !Matches(Target)) && TryGetTarget(ref target) && canSwitch)
                    Target = target;
                break;

            case StickinessType.UntilDeath:
                if (Target == null && TryGetTarget(ref target) && canSwitch)
                    Target = target;
                break;

            case StickinessType.Lifelink:
            case StickinessType.Soulbond:
                if (!EverTargetted && TryGetTarget(ref target) && canSwitch)
                    Target = target;
                else if (Target == null && (EverTargetted || Stickiness == StickinessType.Soulbond))
                    RemoveSelf();
                break;

            default: throw new Exception("unreachable");
        }
        Collidable = false;

        EverTargetted |= Target != null;
    }

    private bool TryGetTarget(ref Entity? target)
        => (target ??= Filter(Candidates)) != null;

    private IEnumerable<Entity> Candidates {
        get {
            switch (WillTarget) {
                case TargetType.Player: return Scene.Tracker.GetEntities<Player>();
                case TargetType.Actor: return Scene.Tracker.GetEntities<Actor>();
                case TargetType.NonPlayerActor: return Scene.Tracker.GetEntities<Actor>().Where(c => c is not Player);
                case TargetType.Solid: return Scene.Tracker.GetEntities<Solid>();
                default: throw new Exception("unreachable");
            }
        }
    }

    private Entity? Filter(IEnumerable<Entity> candidates) {
        switch (Detection) {
            case DetectionType.Within:
            case DetectionType.Intersecting:
                candidates = candidates.Where(Matches);
                goto case DetectionType.Nearest;

            case DetectionType.Nearest:
                return candidates.MinBy(c => (Center - c.Center).LengthSquared());

            default: throw new Exception("unreachable");
        }
    }

    private bool Matches(Entity e) {
        if (e is SolidTiles) return false;

        switch (Detection) {
            case DetectionType.Within: return IsWithin(e);
            case DetectionType.Intersecting: return IsIntersecting(e);
            default: return true;
        }
    }

    private bool IsWithin(Entity e) {
        if (e.Collider == null)
            return Collide.CheckPoint(this, e.Position);

        return (
            e.Collider.AbsoluteLeft >= Left
            && e.Collider.AbsoluteRight <= Right
            && e.Collider.AbsoluteTop >= Top
            && e.Collider.AbsoluteBottom <= Bottom
        );
    }

    private bool IsIntersecting(Entity e) {
        if (e.Collider == null)
            return Collide.CheckPoint(this, e.Position);

        return CollideCheck(e);
    }

    #endregion Targetting
    #region --- Rendering ---

    public override void DebugRender(Camera camera) {
        base.DebugRender(camera);

        var canSwitch = Enabled;
        switch (Stickiness) {
            case StickinessType.Free:
                break;
            case StickinessType.Transient:
            case StickinessType.UntilNewMatch:
                canSwitch &= Target == null || !Matches(Target); break;
            default:
                canSwitch &= Target == null; break;
        }

        Draw.HollowRect(X, Y, Width, Height, canSwitch ? Color.Red : Color.DarkRed);

        if (EverTargetted)
            Draw.Line(Center, new(SliderX.Value, SliderY.Value), Target == null ? Color.DarkRed : Color.Red);
    }

    #endregion Rendering


}

File: Loenn/triggers/PositionTrackerRegion.lua
local utils = require("utils")

local nonEmptyValidator = function(s)
  return s ~= ""
end

local fieldInformation = {
  target = {
    options = {
      "Player",
      "Actor",
      "NonPlayerActor",
      "Solid"
    },
    editable = false
  },
  detection = {
    options = {
      "Within",
      "Intersecting",
      "Nearest"
    },
    editable = false
  },
  stickiness = {
    options = {
      "Free",
      "Transient",
      "UntilNewMatch",
      "UntilDeath",
      "Lifelink",
      "Soulbond"
    },
    editable = false
  },
  tracking = {
    options = {
      "Position",
      "Center",
      "TopCenter",
      "BottomCenter",
      "CenterLeft",
      "CenterRight",
      "Size"
    },
    editable = false
  },
  sliderPrefix = {validator = nonEmptyValidator}
}

local abbreviations = {
  target = {
    Player = "P",
    Actor = "A",
    NonPlayerActor = "N",
    Solid = "S",
  },
  detection = {
    Within = "W",
    Intersecting = "I",
    Nearest = "N",
  },
  stickiness = {
    Free = "F",
    Transient = "T",
    UntilNewMatch = "N",
    UntilDeath = "D",
    Lifelink = "L",
    Soulbond = "S",
  },
  tracking = {
    Position = "P",
    Center = "C",
    TopCenter = "T",
    BottomCenter = "B",
    CenterLeft = "L",
    CenterRight = "R",
    Size = "S",
    Speed = "V",
  },
  directionality = {
    EntityToSliders = ">",
    SlidersToEntity = "<",
    AddDeltas = "+",
  },
}

local fieldOrder = {
  "x",
  "y",
  "width",
  "height",

  "target",
  "detection",
  "stickiness",
  "tracking",
  "directionality",

  "sliderPrefix",
  "targettingFlag",

  "retargetIfFlag",
  "invertRetargetIfFlag",
  "retargetIfExpression",
}

local function abbr(str, options)
  return options[str] or "?"
end

local function triggerTextFlag(room, trigger)
  return (
    "Position Tracker Region (Flag) - "
    .. trigger.sliderPrefix
    .. " ("
    .. abbr(trigger.target or "Player", abbreviations.target)
    .. abbr(trigger.detection or "Intersecting", abbreviations.detection)
    .. abbr(trigger.stickiness or "Lifelink", abbreviations.stickiness)
    .. abbr(trigger.tracking or "Position", abbreviations.tracking)
    .. abbr(trigger.directionality or "EntityToSliders", abbreviations.directionality)
    .. ")"
  )
end

local function triggerTextExpr(room, trigger)
  return (
    "Position Tracker Region (Expression) - "
    .. trigger.sliderPrefix
    .. " ("
    .. abbr(trigger.target or "Player", abbreviations.target)
    .. abbr(trigger.detection or "Intersecting", abbreviations.detection)
    .. abbr(trigger.stickiness or "Lifelink", abbreviations.stickiness)
    .. abbr(trigger.tracking or "Position", abbreviations.tracking)
    .. abbr(trigger.directionality or "EntityToSliders", abbreviations.directionality)
    .. ")"
  )
end

return {
  {
    name = "Microlith57Misc/PositionTrackerRegion",
    placements = {
      {
        name = "positionTrackerRegion",
        data = {
          width = 16,
          height = 16,
          retargetIfFlag = "",
          invertRetargetIfFlag = false,

          target = "Actor",
          detection = "Within",
          stickiness = "Soulbond",
          tracking = "Position",
          directionality = "EntityToSliders",

          sliderPrefix = "trackedPosition",
          targettingFlag = "",
        }
      }
    },
    fieldInformation = fieldInformation,
    fieldOrder = fieldOrder,
    triggerText = triggerTextFlag,
  },
  {
    name = "Microlith57Misc/PositionTrackerRegion_Expression",
    associatedMods = {"Microlith57MiscellaneousMechanics", "FrostHelper"},
    placements = {
      {
        name = "positionTrackerRegion",
        data = {
          width = 16,
          height = 16,
          retargetIfExpression = "",

          target = "Actor",
          detection = "Within",
          stickiness = "Soulbond",
          tracking = "Position",
          directionality = "EntityToSliders",

          sliderPrefix = "trackedPosition",
          targettingFlag = "",
        }
      }
    },
    fieldInformation = fieldInformation,
    fieldOrder = fieldOrder,
    triggerText = triggerTextExpr,
  }
}
